### PR TITLE
Mail/UI Tree: Added session based persistence of node states (similar…

### PR DESCRIPTION
**Pull request by mjansenDatabay**
… to ilExplorerBaseGUI)

Mantis Issue: https://mantis.ilias.de/view.php?id=26055

### CODE REVIEW

File: src/UI/templates/js/Tree/tree.js [hunk 1]
```diff
@@ -4,26 +4,35 @@ il.UI = il.UI || {};
 (function($, UI) {
 	UI.tree = (function($) {
 
-		var init = function (component_id, highlight_nodes) {
+		const toggleNodeState = function (event, action, state_param, was_expanded) {
```
Comment (mjansenDatabay) [inline src/UI/templates/js/Tree/tree.js:7]: @klees / @Amstutz  : Maybe you could review the changes regarding this file.

File: Services/Mail/classes/class.ilMailGUI.php [hunk 2]
```diff
@@ -281,4 +281,29 @@ private function showHeader() : void
...
+            if (0 === $priorState && !in_array($nodeId, $openNodes)) {
+                $openNodes[] = $nodeId;
+                $store->set('mail_tree', serialize($openNodes));
+            } elseif (1 === $priorState && in_array($nodeId, $openNodes)) {
+                $key = array_search($nodeId, $openNodes);
+                unset($openNodes[$key]);
+                $store->set('mail_tree', serialize($openNodes));
+            }
+        }
+        exit();
```
Comment (alex40724) [inline Services/Mail/classes/class.ilMailGUI.php:309]: Consumer code needs to take care of toggle states of the tree?

**Review by klees — APPROVED** [review]
Hi @mjansenDatabay,

thanks a lot! IMO this looks fine for the moment. The code looks fairly low-level for the consumer and I'm sure we can improve the handling of persistent node state in the future. I would be happy if you drop us an according entry in the roadmap that points to your code.

Thanks!

File: Services/Mail/classes/class.ilMailGUI.php (hunk 2 above)
Comment (mjansenDatabay) [reply Services/Mail/classes/class.ilMailGUI.php:309]: Currently we have no central mechanism and controller for this.

For most trees this (or something similar) could be done/implemented in `ilExplorerBaseGUI` or `ilTreeExplorerGUI`.

**Comment by alex40724** [comment]
Consumers (ilMailGUI) must not be responsible to persist tree internal node states information. If UIComponent/Explorer2 should be the official wrapper for the new KS trees (which I did not expect), the mail tree should use this, too, imo.

**Comment by mjansenDatabay** [comment]
> Consumers (ilMailGUI) must not be responsible to persist tree internal node states information. If UIComponent/Explorer2 should be the official wrapper for the new KS trees (which I did not expect), the mail tree should use this, too, imo.

@alex40724 I agree, if there is a centralized wrapper for the new `UI Tree`, the mail system should also use it. But currently there is no such generic wrapper (neither for command/persistence handling, nor for the UI part). Furthermore this wrapper must be aware of different tree implementations (e.g. the `KS Tree` in `Services/Mail` differs from that in `Modules/Forum` in regards of node types, hidden root node etc., so that's why my components cannot use `ilTreeExplorerGUI` anymore).

**Comment by mjansenDatabay** [comment]
@klees I will create a new roadmap entry.

@Amstutz Any objections?

**Comment by alex40724** [comment]
So currently the forum already would need to include similar code that persist tree internal node states, right? Or do I miss something?

I will check how far this would work in UIComponent/Explorer2. But I would rather wait for a decent wrapper before having too many copies of this in the code base.

**Comment by mjansenDatabay** [comment]
@alex40724 Yes, the forum needs to include similar code.

Centralizing this in `ilTreeExplorerGUI` should be possible, if every derived class is able to build it's own node types and tree attributes in an overwritten method (without having to duplicate lots of code of the respective parent method).

**Comment by Amstutz** [comment]
@mjansenDatabay no objections. Thx a lot for your work! Centralizing this for all trees extending ilTreeExplorerGUI would be IMO the right move for the moment. However, this would not help in this case, right? So IMO this is fine as solution for the moment.

In general I agree with @alex40724 that this is not something the client should need to take care of at this low level.

**Comment by mjansenDatabay** [comment]
Will be closed if #2252 is more appreciated (especially by @alex40724 ).

**Pull request closed without merging**